### DISCUSSION TITLE
Cannot find module '@auth/drizzle-adapter'

### DISCUSSION THREAD

**Discussion by davidmytton** [Xunnamius/next-test-api-route-handler#983] [Q&A] [answered]
Thanks for building this library! I'm wondering if there's any setup for Jest & Next.js required that's not documented because the readme jumps straight into installing `next-test-api-route-handler`?

I've followed the guide at https://nextjs.org/docs/app/building-your-application/testing/jest but using `jest-environment-node` instead of `jsdom` which has resulted in this `jest.config.ts`:

```ts
import type { Config } from "jest";
import nextJest from "next/jest.js";

const createJestConfig = nextJest({
  // Provide the path to your Next.js app to load next.config.js and .env files in your test environment
  dir: "./",
});

// Add any custom config to be passed to Jest
const config: Config = {
  coverageProvider: "v8",
  testEnvironment: "node",
  // Add more setup options before each test is run
  // setupFilesAfterEnv: ['<rootDir>/jest.setup.ts'],
};

// createJestConfig is exported this way to ensure that next/jest can load the Next.js config which is async
export default createJestConfig(config);
```

I have a simple test for a `GET` app router API handler:

```ts
import { testApiHandler } from "next-test-api-route-handler";
import * as appHandler from "./route";

it("returns 403 unauthorized when not authenticated", async () => {
  expect.hasAssertions();

  await testApiHandler({
    appHandler,
    test: async ({ fetch }) => {
      await expect((await fetch()).json()).resolves.toStrictEqual({
        error: "Unauthorized",
      });
    },
  });
});
```

But when I run `npx jest` I get an error:

```
  ● Test suite failed to run

    Cannot find module '@auth/drizzle-adapter' from 'lib/auth.ts'

    Require stack:
      lib/auth.ts
      data/auth.ts
      data/sites.ts
      app/api/2023-11-08/sites/route.ts
      app/api/2023-11-08/sites/route.test.ts

      10 |   adapter: DrizzleAdapter(db) as Adapter, // Typecast required in https://github.com/nextauthjs/next-auth/issues/9493
      11 |   providers: [
    > 12 |     process.env.NODE_ENV !== "production"
         |                         ^
      13 |       ? CredentialsProvider({
      14 |           name: "test",
      15 |           credentials: {

      at Resolver._throwModNotFoundError (../../node_modules/jest-resolve/build/resolver.js:427:11)
      at Object.<anonymous> (lib/auth.ts:12:25)
      at Object.<anonymous> (data/auth.ts:11:15)
      at Object.<anonymous> (data/sites.ts:35:15)
      at Object.<anonymous> (app/api/2023-11-08/sites/route.ts:11:16)
      at Object.<anonymous> (app/api/2023-11-08/sites/route.test.ts:6:56)

Test Suites: 1 failed, 1 total
Tests:       0 total
Snapshots:   0 total
Time:        0.244 s
Ran all test suites.
```

`lib/auth.ts` is importing `import { DrizzleAdapter } from "@auth/drizzle-adapter";` which is installed (`0.6.3`).

Have I missed something?

**Comment by davidmytton** [comment]
I was able to solve this by mocking the `lib/auth.ts` library:

```ts
// route.test.ts
import { testApiHandler } from "next-test-api-route-handler";
import * as appHandler from "./route";
import type { Session } from "next-auth";

let mockedSession: Session | null = null;

// Mock authOptions
jest.mock("@/lib/auth", () => ({
  authOptions: {
    adapter: {},
    providers: [],
    callbacks: {},
  },
}));

jest.mock("next-auth/next", () => ({
  getServerSession: jest.fn(() => Promise.resolve(mockedSession)),
}));

afterEach(() => {
  mockedSession = null;
});

it("returns 403 unauthorized when not authenticated", async () => {
  mockedSession = null;

  await testApiHandler({
    appHandler,
    test: async ({ fetch }) => {
      const response = await fetch({ method: "GET" });
      console.log(await response.json());
      expect(response.status).toBe(401);
    },
  });
});
```

And this test now passes:

```
npx jest
  console.log
    { error: 'Unauthorized' }

      at test (app/api/2023-11-08/teams/route.test.ts:31:15)

 PASS  app/api/2023-11-08/teams/route.test.ts
  ✓ returns 403 unauthorized when not authenticated (42 ms)

Test Suites: 1 passed, 1 total
Tests:       1 passed, 1 total
Snapshots:   0 total
Time:        0.44 s, estimated 1 s
Ran all test suites.
```

However, if I attempt to check the returned JSON:

```ts
it("returns 403 unauthorized when not authenticated", async () => {
  mockedSession = null;

  await testApiHandler({
    appHandler,
    test: async ({ fetch }) => {
      const response = await fetch({ method: "GET" });
      console.log(await response.json());
      expect(response.status).toBe(401);
      await expect(response.json()).resolves.toStrictEqual({
        error: "Unauthorized",
      });
    },
  });
});
```

I get an error:

```
npx jest
  console.log
    { error: 'Unauthorized' }

      at test (app/api/2023-11-08/teams/route.test.ts:31:15)

 FAIL  app/api/2023-11-08/teams/route.test.ts
  ✕ returns 403 unauthorized when not authenticated (41 ms)

  ● returns 403 unauthorized when not authenticated

    TypeError: ReactIs.isElement is not a function

      31 |       console.log(await response.json());
      32 |       expect(response.status).toBe(401);
    > 33 |       await expect(response.json()).resolves.toStrictEqual({
         |       ^
      34 |         error: "Unauthorized",
      35 |       });
      36 |     },

      at Object.test (../../node_modules/pretty-format/build/plugins/ReactElement.js:136:44)
      at findPlugin (../../node_modules/pretty-format/build/index.js:302:22)
      at format (../../node_modules/pretty-format/build/index.js:451:22)
      at ../../node_modules/expect/build/index.js:190:46
      at async test (app/api/2023-11-08/teams/route.test.ts:33:7)
      at async testApiHandler (../../node_modules/next-test-api-route-handler/dist/src/index.js:122:5)
      at async Object.<anonymous> (app/api/2023-11-08/teams/route.test.ts:27:3)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        0.254 s, estimated 1 s
Ran all test suites.
```

I have `import "@testing-library/jest-dom";` in `jest.setup.ts` which is included in the `jest.config.ts` in `setupFilesAfterEnv: ["./jest.setup.ts"],`

**Reply by davidmytton** [reply]
This was caused by adding `react: "next/dist/compiled/react/cjs/react.development.js",` to the `moduleNameMapper` to get access to the `cache` function Next.js provides. Mocking that resolved this error.

**Comment by davidmytton** [comment]
Mocking the use of `cache` described in https://github.com/Xunnamius/next-test-api-route-handler/discussions/983#discussioncomment-8405128 I now have this test:

```ts
it("returns 403 unauthorized when not authenticated", async () => {
  mockedSession = null;

  await testApiHandler({
    appHandler,
    test: async ({ fetch }) => {
      const response = await fetch({ method: "GET" });
      console.log(await response.json());
      expect(response.status).toBe(401);
      /*await expect(response.json()).resolves.toStrictEqual({
        error: "Unauthorized",
      });*/
    },
  });
});
```

This works correctly. However, if I uncomment `expect(response.json()).resolves.toStrictEqual` it errors:

```
npm run test

> app@0.1.0 test
> jest

  console.log
    { error: 'Unauthorized' }

      at test (app/api/2023-11-08/teams/route.test.ts:42:15)

 FAIL  app/api/2023-11-08/teams/route.test.ts
  ✕ returns 403 unauthorized when not authenticated (40 ms)

  ● returns 403 unauthorized when not authenticated

    expect(received).resolves.toStrictEqual()

    Received promise rejected instead of resolved
    Rejected to value: [TypeError: Body is unusable]

      42 |       console.log(await response.json());
      43 |       expect(response.status).toBe(401);
    > 44 |       await expect(response.json()).resolves.toStrictEqual({
         |             ^
      45 |         error: "Unauthorized",
      46 |       });
      47 |     },

      at expect (../../node_modules/expect/build/index.js:113:15)
      at test (app/api/2023-11-08/teams/route.test.ts:44:13)
      at async testApiHandler (../../node_modules/next-test-api-route-handler/dist/src/index.js:122:5)
      at async Object.<anonymous> (app/api/2023-11-08/teams/route.test.ts:38:3)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 total
Snapshots:   0 total
Time:        0.346 s, estimated 1 s
Ran all test suites.
```

**Comment by Xunnamius** [comment]
Hi there! NTARH is conceptually quite simple: it's a thin wrapper around `node:http`'s `createServer` that passes the request and response to an internal Next.js function and returns the result. This means you can even run NTARH entirely outside of Jest or any testing framework so long as you have Next.js installed. It is for that reason that there isn't really any additional guidance for Jest or Vitest or other frameworks since NTARH operates orthogonally to them.

That last error is interesting though, ~~that might be due to NTARH working around the mismatch between Node streams and WHATWG streams. Do you have a dummy repo that minimally reproduces this error? If so, I'll take a look.~~ Whoops, I skimmed your code too quickly, my bad. When you call `console.log(await response.json());`, that [consumes the body](https://fetch.spec.whatwg.org/#concept-body-consume-body). You cannot later call `await expect(response.json())` on the same request since the body has already been consumed :) If you comment out that first line before uncommenting that second line, do you still get the error?

**Reply by davidmytton** [reply]
Yes, that was it! Left that in whilst debugging the response. The `GET` tests now work, but when I send a `DELETE` or `POST` request I get:

```
 expect(received).resolves.toStrictEqual(expected) // deep equality

    Expected: {"error": "Unauthorized"}
    Received: "TypeError: RequestInit: duplex option is required when sending a body.
        at new Request (node:internal/deps/undici/undici:5499:19)
        at normalizeNodeRequest (/Users/david/Code/a/node_modules/@whatwg-node/server/cjs/utils.js:192:12)
        at handleNodeRequest (/Users/david/Code/a/node_modules/@whatwg-node/server/cjs/createServerAdapter.js:104:61)
        at requestListener (/Users/david/Code/a/node_modules/@whatwg-node/server/cjs/createServerAdapter.js:118:25)
        at genericRequestHandler (/Users/david/Code/a/node_modules/@whatwg-node/server/cjs/createServerAdapter.js:229:20)
        at Object.apply (/Users/david/Code/a/node_modules/@whatwg-node/server/cjs/createServerAdapter.js:298:20)
        at Server.<anonymous> (/Users/david/Code/a/node_modules/next-test-api-route-handler/dist/src/index.js:235:9)
        at Server.emit (node:events:517:28)
        at Server.emit (node:domain:489:12)
        at parserOnIncoming (node:_http_server:1123:12)
        at HTTPParser.parserOnHeadersComplete (node:_http_common:119:17)"

      45 |       //console.log(await response.json());
      46 |       //expect(response.status).toBe(401);
    > 47 |       await expect(response.text()).resolves.toStrictEqual({
         |                                              ^
      48 |         error: "Unauthorized",
      49 |       });
      50 |     },

      at Object.toStrictEqual (../../node_modules/expect/build/index.js:174:22)
      at test (app/api/2023-11-08/teams/[id]/route.test.ts:47:46)
      at async testApiHandler (../../node_modules/next-test-api-route-handler/dist/src/index.js:122:5)
      at async Object.<anonymous> (app/api/2023-11-08/teams/[id]/route.test.ts:38:3)

 PASS  app/api/2023-11-08/teams/route.test.ts
```

The code looks like this:

```ts
it("returns 403 unauthorized when not authenticated", async () => {
  mockedSession = null;

  await testApiHandler({
    appHandler,
    test: async ({ fetch }) => {
      const response = await fetch({
        method: "DELETE",
        body: "test",
      });
      //console.log(await response.json());
      //expect(response.status).toBe(401);
      await expect(response.text()).resolves.toStrictEqual({
        error: "Unauthorized",
      });
    },
  });
});
```

**Reply by Xunnamius** [reply]
That's one of those interesting results of an [impedance mismatch](https://github.com/whatwg/fetch/issues/1254) between the fetch spec, implementers of the fetch spec, and other implementers of the fetch spec.

(what comes next is only AFAIK, please feel free to correct me; I'll leave this explanation here to link back to later on if the README needs updating)

Currently, when calling fetch with `const res = await fetch(...)`, (#‎1) the request body is streamed to the server to completion, then (#‎2) the server returns response headers which causes `const res = await fetch(...)` to resolve. The server then (#‎3) finishes streaming the response body, which is why we have to `await` again when we call `await res.json()` or `await res.text()`. This is "half duplex" streaming where data is only ever traveling in one direction across the pipe at a time.

Node.js (and Deno, Ruby, Python, and others) technically supports "full duplex" streams for the fetch body, where the server and client both send data to each other in parallel (scratching out #‎1 above), which would allow `const res = await fetch(...)` to resolve before the request body stream is closed / streamed to completion. The HTTP spec also allows this, but the WHATWG fetch spec is apparently ambivalent given some error/edge cases. Therefore, it is the case that no browsers currently implement "full duplex" body streaming AFAIK. But everyone rightly wants to leave room for when it'll eventually land, maybe tomorrow, maybe a couple decades from now.

Hence the `Received: "TypeError: RequestInit: duplex option is required when sending a body.` line in your error message. To fix this, when calling `fetch` from Node.js land while streaming, you'll need to tell Node.js [explicitly](https://github.com/whatwg/fetch/pull/1457) how you want to handle data streaming. Since the fetch spec only technically supports the "half duplex" mode right now, your only option is "half," i.e. `await fetch(..., { duplex: 'half' })`. Maybe later you could also choose "full" (I think Deno supports "full" right now). [Understandably](https://github.com/nodejs/node/issues/46221), it'd be potentially dangerous for most fetch implementers to choose a default for you given the future prospect of an entirely new streaming mode that might break your assumptions. So for the purposes of future proofing, they want you to be explicit.

Try:

```typescript
it("returns 403 unauthorized when not authenticated", async () => {
  mockedSession = null;

  await testApiHandler({
    appHandler,
    test: async ({ fetch }) => {
      const response = await fetch({
        method: "DELETE",
        body: "test", // <== this is a bit weird though since this isn't a stream
        duplex: "half"
      });
      //console.log(await response.json());
      //expect(response.status).toBe(401);
      await expect(response.text()).resolves.toStrictEqual({
        error: "Unauthorized",
      });
    },
  });
});
```

Unfortunately, the Node.js docs are lacking here since they just redirect to the MDN docs.


Though, you shouldn't be getting this error when the request body isn't a stream, so you may need to add `duplex: 'half'` inside your `appHandler` if you're doing additional fetching there. If that still isn't working, it may be an NTARH issue, so let me know!

**Reply by davidmytton** [reply]
Thanks for the detailed explanation. Unfortunately that didn't help. Adding `duplex` causes VS Code to complain with `Object literal may only specify known properties, and 'duplex' does not exist in type 'RequestInit'.ts(2353)` and I get the same error when running the test:

```
      TypeError: RequestInit: duplex option is required when sending a body.
          at new Request (node:internal/deps/undici/undici:5499:19)
          at normalizeNodeRequest (/Users/david/Code/a/node_modules/@whatwg-node/server/cjs/utils.js:192:12)
          at handleNodeRequest (/Users/david/Code/a/node_modules/@whatwg-node/server/cjs/createServerAdapter.js:104:61)
          at requestListener (/Users/david/Code/a/node_modules/@whatwg-node/server/cjs/createServerAdapter.js:118:25)
          at genericRequestHandler (/Users/david/Code/a/node_modules/@whatwg-node/server/cjs/createServerAdapter.js:229:20)
          at Object.apply (/Users/david/Code/a/node_modules/@whatwg-node/server/cjs/createServerAdapter.js:298:20)
          at Server.<anonymous> (/Users/david/Code/a/node_modules/next-test-api-route-handler/dist/src/index.js:235:9)
          at Server.emit (node:events:517:28)
          at Server.emit (node:domain:489:12)
          at parserOnIncoming (node:_http_server:1123:12)
          at HTTPParser.parserOnHeadersComplete (node:_http_common:119:17)
```

This is NodeJS v18.19.0.

**Reply by Xunnamius** [reply]
Yeah, the TypeScript types for Node.js have not caught up yet last I checked. Still, you shouldn't have to use that option if you're not passing a stream for body. Can you link a minimal repo that demonstrates your error? I'll take a look later today.

**Reply by davidmytton** [reply]
I've created a simple repo at https://github.com/davidmytton/ntarh-test which has a failing test. This isn't the same as I reported above, but is failing when I think it should be passing...

```
npm run test

> api-tests@0.1.0 test
> jest

 FAIL  app/api/test/route.test.ts
  ✓ GET returns 200 (33 ms)
  ✕ POST returns 200 (4 ms)

  ● POST returns 200

    TypeError: Request with GET/HEAD method cannot have body.

      at new NextRequest (node_modules/next/src/server/web/spec-extension/request.ts:24:10)
      at node_modules/next-test-api-route-handler/dist/src/index.js:167:57
      at handleNodeRequest (node_modules/@whatwg-node/server/cjs/createServerAdapter.js:105:16)
      at requestListener (node_modules/@whatwg-node/server/cjs/createServerAdapter.js:118:25)
      at genericRequestHandler (node_modules/@whatwg-node/server/cjs/createServerAdapter.js:229:20)
      at Object.apply (node_modules/@whatwg-node/server/cjs/createServerAdapter.js:298:20)
      at Server.<anonymous> (node_modules/next-test-api-route-handler/dist/src/index.js:235:9)

Test Suites: 1 failed, 1 total
Tests:       1 failed, 1 passed, 2 total
Snapshots:   0 total
Time:        0.302 s, estimated 1 s
Ran all test suites.
```

**Reply by Xunnamius** [reply]
~~Aha! This issue is related to jsdom. NTARH is currently only being tested on `{ testEnvironment: 'node' }` instead of `{ testEnvironment: 'jsdom' }`, the latter providing a mock `Request` class that doesn't spread the same way Node's native `Request` does. I'll investigate a fix.~~ Ah, I spoke too soon. The spreading problem is the issue, and fixing it fixes your tests, but I'm still investigating why the unit tests pass while your demo repo's second test does not pass; it isn't jsdom related (I was looking at the wrong config file). 

UPDATE: When I replace SWC with babel and remove the `jest.config.ts` file, all your tests pass. So it seems to be SWC-related and/or related to the jest configuration generated by `next/jest.js`. I'll be investigating this further.

Thank you for reporting this issue!

**Comment by Xunnamius** [comment, marked as answer]
This has been addressed in [v4.0.5](https://github.com/Xunnamius/next-test-api-route-handler/releases/tag/v4.0.5).

I'll be using your demo repo as an integration test for `next/jest.js` and SWC going forward :) If the latest version of NTARH still isn't working for you, please open a new issue and I'll investigate.

**Reply by davidmytton** [reply]
Great, thanks. Confirmed v4.0.5 fixes this!
